### PR TITLE
Step Packages Cycle 2 Braindump

### PR DESCRIPTION
**Pull request by andrewabest**
A braindump of the current state of Step Packages, collating learnings and decisions from Cycle 2 2021.

### CODE REVIEW

File: Steps/Components/StepPackages.md [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+Dotnet will only be used for migrating existing Sashimi-based steps into Step Packages.
+
+![Step Packages](https://github.com/OctopusDeploy/Architecture/blob/master/Steps/assets/building_blocks.png)
+
+## Conventions
+
+Step Packages use a convention-based structure:
+
+```
+  |-- package.json
```
Comment (johnsimons) [inline Steps/Components/StepPackages.md:24]: What is this?

File: Steps/Components/StepPackages.md [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+  |-- other node tooling configuration files
+  |-- stepA/executor.ts
+  |-- stepA/inputs.ts
+  |-- stepA/logo.svg
```
Comment (johnsimons) [inline Steps/Components/StepPackages.md:28]: A logo can be svg or any other image format

File: Steps/Components/StepPackages.md [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+
+![Step Packages](https://github.com/OctopusDeploy/Architecture/blob/master/Steps/assets/building_blocks.png)
+
+## Conventions
+
+Step Packages use a convention-based structure:
+
+```
+  |-- package.json
+  |-- other node tooling configuration files
```
Comment (johnsimons) [inline Steps/Components/StepPackages.md:25]: What is this?

File: Steps/Components/StepPackages.md [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+
+**Step Packages** are Octopus Server's extensibility model for creating new Steps - the things that do the work of deployment!
+
+A Step Package is a `zip` file. It contains all of the things Octopus Server needs to use the Step in a Deployment Process.
+
+Step Packages are named in the format `Octopus.MyStepPackage.1.0.0.zip`, and are located under `%OctopusServerRoot%/bin/steps`
+
+Step packages have been designed to allow using different technologies to develop the Step Executor they contain. Currently we support writing Step Packages in node, and will soon support writing them in dotnet.
+
+Node will be our default choice for developing Step Packages.
```
Comment (johnsimons) [inline Steps/Components/StepPackages.md:13]: I don't entirely agree with this, I thought we use the right tool for the right job, not sure node is it

File: Steps/Concepts/Execution.md [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+[Execution Manifest Diagram](https://whimsical.com/steps-execution-manifest-N74pfHgDoLXNK8ck1UJmb9)
+
+[Further Discussion](https://docs.google.com/document/d/1E5u3BnYlLzXQ4kwbQVb6XY9TmSFegxtAnwBi_A77WAE)
+
+# Step Execution
+
+Steps from Step Packages are executed by [Calamari](https://github.com/octopusdeploy/Calamari).
+
+Calamari, upon recieving an `execute-manifest` command with an accompanying Execution Manifest and variables, will do the work of marshalling the execution of the step.
```
Comment (johnsimons) [inline Steps/Concepts/Execution.md:79]: spelling

File: Steps/Components/StepPackages.md (hunk 2 above)
Comment (andrewabest) [reply Steps/Components/StepPackages.md:28]: I don't think it can be at the moment - the logo controller responds with a hard-coded mime type in Server. Perhaps that doesn't impact this.

File: Steps/Components/StepPackages.md (hunk 3 above)
Comment (andrewabest) [reply Steps/Components/StepPackages.md:25]: Perhaps that confuses things, I was trying to make it clear that aside from the files required there by convention, you'd also have your usual node files like `package.json` etc in your step package.

File: Steps/Components/StepPackages.md (hunk 4 above)
Comment (andrewabest) [reply Steps/Components/StepPackages.md:13]: The choice of `node` as the default reflects our intention for this team - we want to grow it primarily with node developers. By having this as a default we lean into that intention. The other reason for a default is just consistency - by having most of our steps be the same shape, they will be easier to maintain over time. This doesn't mean we wouldn't use dotnet if it were clearly a better choice for some reason, but it isn't a choice we would look to make for every new step package - we would default to node, unless there was a compelling reason to use dotnet.

File: Steps/Components/StepPackages.md (hunk 1 above)
Comment (andrewabest) [reply Steps/Components/StepPackages.md]: I've removed it

File: Steps/Components/StepPackages.md (hunk 3 above)
Comment (andrewabest) [reply Steps/Components/StepPackages.md]: I've removed it

File: Steps/Concepts/Execution.md (hunk 5 above)
Comment (andrewabest) [reply Steps/Concepts/Execution.md]: Done

File: Steps/Components/StepPackages.md [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+
+Dotnet will only be used for migrating existing Sashimi-based steps into Step Packages.
+
+![Step Packages](https://github.com/OctopusDeploy/Architecture/blob/master/Steps/assets/building_blocks.png)
+
+## Conventions
+
+Step Packages use a convention-based structure:
```
Comment (TomPeters) [inline Steps/Components/StepPackages.md:21]: I think there's some language confusion here (it tripped my up when I was reading this)

- Step Packages: The compiled and packaged zip file that is consumed by Octopus Server
- Step Package Repo / Step Package Source: The source files and contents of the repository which are compiled and packaged by the Step Package CLI.

This section seems to be talking about the latter, but the language confused me and made me think you were talking about the former.

Documenting both are useful, depending on whether someone is trying to understand how to author a  step package or how to consume one in Octopus Server.

File: Steps/Components/StepPackages.md [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+
+Step packages have been designed to allow using different technologies to develop the Step Executor they contain. Currently we support writing Step Packages in node, and will soon support writing them in dotnet.
+
+Node will be our default choice for developing Step Packages.
+
+Dotnet will only be used for migrating existing Sashimi-based steps into Step Packages.
+
+![Step Packages](https://github.com/OctopusDeploy/Architecture/blob/master/Steps/assets/building_blocks.png)
+
+## Conventions
```
Comment (TomPeters) [inline Steps/Components/StepPackages.md:19]: I think a lot of this detail is too much for the architecture repo.

I think I would expect the architecture repo to contain design details like "We are using conventions for the file structure of the step packages, and this is why" and "We are using a metadata.json file because it is language agnostic and can be easily parsed in any language."

I think specific details about the structure of the step api and how to implement a step would be better placed in documentation in the step-api repository, which this document could link to.

Comment (TomPeters) [reply Steps/Components/StepPackages.md:19]: Said another way: as a step author I probably wouldn't think to go looking in the architecture repo for documentation about the step api.

File: Steps/BoundVariables.md [hunk 8]
```diff
@@ -1,60 +0,0 @@
-# Bound Variables
```
Comment (TomPeters) [inline Steps/BoundVariables.md:1]: Why was this deleted? It is almost entirely still accurate, and I think this is critical information to capture at this level.

File: Steps/Concepts/Versioning.md [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+
+We have a step like `step-package-azurestorage.1.0.0` which is compiled against `step-ui-api.1.0.0`. We then make a breaking change to the `step-ui-api` (and release a new version like `step-ui-api.2.0.0`) and update our framework in portal as well, so that `step-package-azurestorage.1.0.0` is no longer directly compatible with portal. We can either upgrade the `step-ui-api` dependency and release a new version of our step (say `step-package-azurestorage.2.0.0`) to make it compatible and deprecate the old step somehow, or we could have a compatibility shim in portal so that portal can support steps compiled against both `step-ui-api.1.0.0` and `step-ui-api.2.0.0`.
+
+# Versioning
+
+A Step Package is a versioned component. It's version is denoted in its file name - `StepPackage.2.0.0.zip` would denote version `2.0.0` of the Step Package. The version will follow [SemVer 2.0.0](https://semver.org/) versioning semantics.
```
Comment (TomPeters) [inline Steps/Concepts/Versioning.md:41]: >The version will follow [SemVer 2.0.0](https://semver.org/) versioning semantics.

I'm unclear about what this means. 

Specifically, semver talks about compatibility. But we've established further down that there's multiple ways that compatibility could break. 

Are we talking about compatibility with all of them? For example, if you update the step to reference a new major version of the step api, would the step's major version also need to increment? Does this imply that users are forced to opt-in to a new version in the UI, even though the change itself might have been trivial or not user-facing?

File: Steps/Concepts/Versioning.md [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+
+Over time, we may make changes to the contents of Step Packages, and increase their version number accordingly.
+
+**Minor Version Increments:** Within Server, we will automatically retrieve the _latest version_ of the same major version a step was created with when working within any of the steps components - UI, validator, executor, etc. This means for all backwards-compatible feature additions and bug fixes, users will recieve these benefits immediately in their existing processes, without having to update anything.
+
+**Major Version Increments:** For major version increments, a user will need to opt-in to them. There will be a version selector presented on the step UI if a newer version is available. If it is, and the user selects the next version, the step will be upgraded to the selected version. This will not be a reversable process.
```
Comment (TomPeters) [inline Steps/Concepts/Versioning.md:47]: >For major version increments, a user will need to opt-in to them

I think I missed this discussion/decision, and I'm not sure I agree :) I would also love to play devil's advocate in such a discussion.

File: Steps/Concepts/Versioning.md [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+_Versioning_ of a Step Package is about enforcing _internal consistency_ of Step Package components as they are used within Octopus Server - making sure we use the right Step Package components to edit, validate, and execute a step after it has been created, and defining useful tolerances within version ranges that allow us to more easily ship bug fixes and small changes to steps without requiring user intervention to upgrade steps in existing processes.
+
+_Compatibility_ refers to the compatibility of various Step Package components with Server components. There are many compatibility surfaces, which we will explore below.
+
+## Tl;Dr:
+
+- We care about **versions** of Step Packages, and the **compatibility** of their components with Server components.
+- We would like to pursue a monorepo for the API (agree strongly with this) - we are already part way there since we have merged them into a single repo.
+- Compatibility surfaces include: manifest + conventions, UI API, input processing (both server and UI), validation (ditto), execution API.
+- We would like to embed version information for all compatibility surfaces into metadata.json (probably a job for the Step Package CLI
```
Comment (TomPeters) [inline Steps/Concepts/Versioning.md:25]:
```suggestion
- We would like to embed version information for all compatibility surfaces into metadata.json (probably a job for the Step Package CLI)
```

File: Steps/BoundVariables.md (hunk 8 above)
Comment (andrewabest) [reply Steps/BoundVariables.md:1]: It was moved to the inputs file, as I thought it should be grouped under that as it relates to inputs.

Comment (TomPeters) [reply Steps/BoundVariables.md:1]: Ah nice, I just missed that. :nothingtoseehere:

File: Steps/Components/StepPackages.md (hunk 7 above)
Comment (andrewabest) [reply Steps/Components/StepPackages.md:19]: I agree, but I'm not sure where else to put it right now. It could live under the `step-api` repository I suppose - that might make sense. What do you think?

I deliberately wanted to capture some of this detail, as it makes for great context going forward.

I could then gut a lot of this and just summarise key decisions.

File: Steps/Concepts/Versioning.md (hunk 9 above)
Comment (andrewabest) [reply Steps/Concepts/Versioning.md:41]: > if you update the step to reference a new major version of the step api, would the step's major version also need to increment

File: Steps/Concepts/Versioning.md (hunk 10 above)
Comment (andrewabest) [reply Steps/Concepts/Versioning.md:47]: Zooming back up to the top of this file

> This section is detailling our intentions for versioning and compatibility for Step Packages and Octopus Server, but may change as we implement them.

Comment (andrewabest) [reply Steps/Concepts/Versioning.md:47]: Consider it a strawman point at this point :)

File: Steps/Components/StepPackages.md (hunk 7 above)
Comment (TomPeters) [reply Steps/Components/StepPackages.md:19]: I think putting it in the step-api repo makes a lot of sense - I can't think of a better place.

File: Steps/Concepts/Versioning.md (hunk 9 above)
Comment (TomPeters) [reply Steps/Concepts/Versioning.md:41]: I think adding that extra detail here is valuable :+1:

File: Steps/Components/StepPackages.md (hunk 7 above)
Comment (andrewabest) [reply Steps/Components/StepPackages.md]: Moved a bunch of the detail to https://github.com/OctopusDeploy/step-api

File: Steps/Concepts/Versioning.md (hunk 9 above)
Comment (andrewabest) [reply Steps/Concepts/Versioning.md]: Done

File: Steps/Components/StepPackages.md (hunk 6 above)
Comment (andrewabest) [reply Steps/Components/StepPackages.md]: I think splitting out the detail has helped this.
